rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,476 @@
+This would allow partners to create an unstyled or custom library easily.
+Partners could re-run the tool to keep up with new components shipping in react-components.
+
+```
+yarn create-recomposed-library my-components
+```
+
+### 👍
+
+- doesn't add yet another way to customize styles to the architecture
+- uses the hooks composition model for its intended purpose, to handle special cases of customization
+- partners get complete control over customization (vs. limitations that may come from a style-only customizer)
+- allows replacement of Griffel with a different styling system
+- supports tree-shaking out the hooks that are not used
+- follows out recommended dependency abstraction approach for partners with large and complex code bases to make future migration easier.
+
+### 👎
+
+- does not provide partners with an out-of-the-box solution
+- non-trivial effort to create a library with every component re-exported
+- maintenance cost keeping up with new and changed components
+- our style hooks are currently monolithic. We may need to create building block style hooks designed for re-use
+- does not affect (reusable) components which are implemented outside the application's source.
+- no consistency between approaches partners might take in creating centralized custom styling solutions
+
+## Option C: Component level CSS custom properties
+
+### 💡 Improvement - Remove dependency on classNames and DOM order
+
+```tsx
+// Button with component tokens and global fallbacks
+const ButtonTokens = {
+  background: '--fui-button-background',
+  color: '--fui-button-color',
+  border: '--fui-button-complex-selector',
+};
+const useButtonStyles = makeResetStyles({
+  backgroundColor: `var(${ButtonTokens.background}, ${tokens.colorBrandBackground})`,
+  color: `var(${ButtonTokens.color}, ${tokens.colorNeutralForegroundOnBrand})`,
+  border: '4px solid',
+  ':enabled:not(:checked):not(:indeterminate)': {
+    borderColor: `var(${ButtonTokens.border}, orange)`,
+  },
+});
+
+const Button = props => {
+  const styles = useButtonStyles();
+  return (
+    <button {...props} className={mergeClasses(styles, props.className)}>
+      {props.children}
+    </button>
+  );
+};
+```
+
+```tsx
+// Customizing button via className
+const useCustomButtonStyle = makeResetStyles({
+  [ButtonTokens.background]: 'red',
+  [ButtonTokens.color]: 'white',
+  [ButtonTokens.border]: 'green',
+  ':hover': {
+    [ButtonTokens.background]: 'green',
+    [ButtonTokens.color]: 'pink',
+    [ButtonTokens.border]: 'blue',
+  },
+  ':active': {
+    [ButtonTokens.background]: 'orange',
+    [ButtonTokens.color]: 'black',
+    [ButtonTokens.border]: 'purple',
+  },
+});
+
+export const CustomButton = props => {
+  const rootStyle = useCustomButtonStyle();
+  return <Button className={rootStyle}>Hello</Button>;
+};
+```
+
+```tsx
+// making changes in composition
+const useCustomButtonStyle = makeResetStyles({
```

Comment:
[You can't do multiple `makeResetStyles()` calls for a single element](https://griffel.js.org/react/api/make-reset-styles#limitations): something must be defined with `makeStyles()` instead.